recipe-server/client/control/utils/handleError.js

```diff
@@ -0,0 +1,84 @@
+import { message as AntMessage } from 'antd';
+
+import APIClient from 'control/utils/api';
+import { ValidationError } from 'control/utils/forms';
+
+
+export const ERR_MESSAGES = {
+  FORM_VALIDATION: 'Please correct the form items highlighted below.',
+  NO_INTERNET: 'Check your internet connection and try again.',
+  SERVER_FAILED: 'The server failed to respond. Please try again.',
+  NOT_LOGGED_IN: 'No active user session found, try logging in again.',
+  NO_PERMISSION: 'You do not have permission to perform that action.',
+};
+
+// Search strings used to determine various types responses.
+const checkFetchFailure = ({ message = '' }) =>
+  message.indexOf('Failed to fetch') !== -1 || message.indexOf('NetworkError') !== -1;
+
+const checkLoginFailure = ({ message = '' }) => message.indexOf('credentials were not provided') > -1;
+const checkAPIFailure = error => error instanceof APIClient.APIError;
+const checkValidationFailure = error => error instanceof ValidationError;
+
+const msgDisplayTime = 8; // seconds
+
+const defaultMethods = {
+  checkUserOnline: () => navigator.onLine,
+  notifyUser: errMsg => AntMessage.error(errMsg, msgDisplayTime),
+};
+
+const handleAPIError = error => {
+  let message = '';
+
+  switch (error.data.status) {
+    case 400: // Bad Request
+      message = ERR_MESSAGES.FORM_VALIDATION;
+      break;
+
+    case 403: // Forbidden
+      message = checkLoginFailure(error) ? ERR_MESSAGES.NOT_LOGGED_IN : ERR_MESSAGES.NO_PERMISSION;
```

Comment:
Technically 403 only applies to a user that is logged in but not authorized to perform an action, whereas 401 is for users that aren't logged in.

We don't do this correctly in dev environments, and server 403 in all situations, which isn't good. I imagine we might serve 401s in prod some times though. We should probably include `case 401:` here as well.

Reply:
Should the `checkLoginFailure` remain, or should this suffice:
```
    case 401: // Not logged in
      message = ERR_MESSAGES.NOT_LOGGED_IN;
      break;

    case 403: // Forbidden
      message = ERR_MESSAGES.NO_PERMISSION;
      break;
```

Reply:
That bit of code is technically correct, but not compatible with Normandy's server, unfortunately. I think that keeping the call to `checkLoginFailure` in the 403 branch, and also including the 401 branch would work better.